internal/hcl/modules/fetch.go

```diff
@@ -95,3 +103,78 @@ func (r *PackageFetcher) fetch(moduleAddr string, dest string) error {
 
 	return nil
 }
+
+// CustomGitGetter extends the standard GitGetter transforming SSH sources to
+// HTTPs first before attempting a Get. This means that we can attempt to use any
+// Git credentials on the host machine to resolve the Get before falling back to
+// SSH.
+type CustomGitGetter struct {
+	*getter.GitGetter
+}
+
+// Get overrides the standard Get method transforming SSH urls to their HTTPS
+// equivalent. Get then tries to get the new url into the dst, falling back to
+// the original SSH url if an HTTPS get fails.
+func (g *CustomGitGetter) Get(dst string, u *url.URL) error {
+	if u.Scheme != "ssh" {
+		return g.GitGetter.Get(dst, u)
+	}
+
+	httpsURL, err := TransformSSHToHttps(u)
+	if err != nil {
+		logging.Logger.Debug().Err(err).Msgf("failed to transform %s to https", u)
+		return g.GitGetter.Get(dst, u)
+	}
+
+	err = g.GitGetter.Get(dst, httpsURL)
+	if err != nil {
+		logging.Logger.Debug().Err(err).Msgf("failed to get transformed ssh url %s, retrying with ssh", httpsURL)
+		return g.GitGetter.Get(dst, u)
+	}
+
+	return nil
+}
+
+// IsGitSSHSource returns if the url u is a valid git ssh source. Param u is
+// expected to be an url that has been transformed by a go-getter Detect pass,
+// removing shorthand and aliased for various sources.
+func IsGitSSHSource(u *url.URL) bool {
+	if u == nil {
+		return false
+	}
+
+	if u.Scheme == "ssh" || u.Scheme == "git::ssh" {
+		return true
+	}
+
+	return false
+}
+
+// TransformSSHToHttps transforms a Terraform module source url to an HTTPS
+// equivalent. This only handles source urls prefixed with ssh:: or git::ssh. The
+// shorthand ssh source referenced here:
+// https://developer.hashicorp.com/terraform/language/modules/sources#github e.g.
+// "git@github.com:hashicorp/example.git" in not handled by this method as we
+// expect the source to already be Detected to the valid longhand equivalent
+// before calling this function. This can be achieved by calling
+// getter.Detect(src) before calling TransformSSHToHttps.
```

Comment:
Just to check:
the `getter.Detect` function handles scp-style addresses like `git::git@example.com/foo/bar` as well?
Also do we handle `git::https://example.com/foo/bar` style addresses?

Reply:
`git::https://example.com/foo/bar` -> yes we handle this. 
`git::git@example.com/foo/bar` -> it seems that this goes to a file loader instead. I'll see if I can make a change to support this

Reply:
sorry this was my confusion - yes the `Dectect` functionality correctly identifies a scp-style address, e.g: as a ssh one: `git@github.com:hashicorp/example.git` -> `git::ssh://git@github.com/hashicorp/example.git` which is expected